storage.go

```diff
@@ -0,0 +1,355 @@
+package tprstorage
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/microstorage"
+	"github.com/giantswarm/operatorkit/tpr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apismeta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	api "k8s.io/client-go/pkg/api/v1"
+)
+
+type TPRConfig struct {
+	Name, Version, Description string
+}
+
+type TPOConfig struct {
+	Name, Namespace string
+}
+
+type Config struct {
+	// Dependencies.
+
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+
+	// Settings.
+
+	// TPR it the third party resource where data objects are stored.
+	TPR TPRConfig
+
+	// TPOName is the third party object used to store data. This
+	// object will be creatd inside a third party resource specified by
+	// TPR. If the object already exists it will be reused. It is
+	// safe to run multiple Storage instances using the same TPO.
+	TPO TPOConfig
+}
+
+func DefaultConfig() Config {
+	return Config{
+		// Dependencies.
+
+		K8sClient: nil, // Required.
+		Logger:    nil, // Required.
+
+		// Settings.
+
+		TPR: TPRConfig{
+			Name:        "tpr-storage.giantswarm.io",
+			Version:     "v1",
+			Description: "Storage data managed by github.com/giantswarm/tprstorage",
+		},
+
+		TPO: TPOConfig{
+			Name:      "", // Required.
+			Namespace: "giantswarm",
+		},
+	}
+}
+
+type Storage struct {
+	logger micrologger.Logger
+	logctx []interface{}
+
+	k8sClient kubernetes.Interface
+	tpr       *tpr.TPR
+
+	tpoEndpoint     string
+	tpoListEndpoint string
+}
+
+func New(ctx context.Context, config Config) (*Storage, error) {
```

Comment:
I know what you mean. But is `Boot()` invoked during event, task or request? I don't think so.

In this case `New` w/o `Boot` wouldn't make sense, and user would have to invoke it right after `New`.

IMO this is only giving user a chance to forget calling `Boot`. I'm trying to be catmatic here.